File: PHP/One/Ast/Interfaces.php
<?php

namespace One\Ast\Interfaces;

interface IType {
    function repr();
}

interface IExpression {
    function setActualType($actualType, $allowVoid, $allowGeneric);
    
    function setExpectedType($type, $allowVoid);
    
    function getType();
}

File: CSharp/One/Ast/Interfaces.cs


namespace One.Ast
{
    public interface IType {
        string repr();
    }
    
    public interface IExpression {
        void setActualType(IType actualType, bool allowVoid, bool allowGeneric);
        
        void setExpectedType(IType type, bool allowVoid);
        
        IType getType();
    }
}